# KNN Classifier atau Klasifikasi pada Data Iris

**Nama  : Oki Bagus Rahmat Prakoso**

**NIM   : 210411100183**

**Kelas : Penambangan Data-C**

Metode K-Nearest Neighbors (KNN) adalah sebuah metode dalam machine learning yang digunakan untuk melakukan klasifikasi atau regresi berdasarkan pada data yang terdekat dengan data yang sedang diprediksi.

Prinsip dasar dari metode KNN adalah mencari sejumlah tetangga terdekat (nearest neighbors) dari data yang sedang diprediksi. Kemudian, label atau nilai target dari tetangga terdekat tersebut digunakan untuk memprediksi label atau nilai target dari data yang sedang diprediksi.

1. Mengimport library yang diperlukan

In [1]:
import pandas as pd
import numpy as np
from sklearn.neighbors import KNeighborsRegressor 
from sklearn.preprocessing import LabelEncoder

2. Import dataset Iris

In [2]:
df = pd.read_csv('https://gist.githubusercontent.com/okiokib/f40e40e4f076d1dbfcae8775be0c118e/raw/71572b2787bf31f6cdb65905afd8832a74db4c0c/irisDataset.csv')

jumlah_baris = df.shape[0]
df.head(jumlah_baris)

sepal_length  sepal_width  petal_length  petal_width    species
0             5.1          3.5           1.4          0.2     setosa
1             4.9          3.0           1.4          0.2     setosa
2             4.7          3.2           1.3          0.2     setosa
3             4.6          3.1           1.5          0.2     setosa
4             5.0          3.6           1.4          0.2     setosa
..            ...          ...           ...          ...        ...
145           6.7          3.0           5.2          2.3  virginica
146           6.3          2.5           5.0          1.9  virginica
147           6.5          3.0           5.2          2.0  virginica
148           6.2          3.4           5.4          2.3  virginica
149           5.9          3.0           5.1          1.8  virginica

[150 rows x 5 columns]

3. Mengecek nilai yang hilang/kosong pada data

In [3]:
df_copy = df.copy()
df_copy.isnull().sum()

sepal_length    0
sepal_width     0
petal_length    0
petal_width     0
species         0
dtype: int64

4. Fungsi dari df_copy = df_copy.replace('?', np.nan) adalah untuk menggantikan nilai yang memiliki tanda tanya ("?") dalam DataFrame df_copy dengan nilai NaN (Not a Number) menggunakan fungsi replace() dari pandas.

>Pada code tersebut, '?' merupakan nilai yang ingin diganti, sedangkan np.nan adalah representasi nilai NaN yang disediakan oleh modul numpy. Nilai NaN digunakan untuk menandakan nilai yang hilang atau tidak valid dalam dataset.

>kesimpulannya untuk mengganti data yang hilang

In [4]:
df_copy = df_copy.replace('?',np.nan)

5. Cek lagi berapa nilai yang kosong pada data

In [5]:
df_copy.isnull().sum()

sepal_length    0
sepal_width     0
petal_length    0
petal_width     0
species         0
dtype: int64

6. Setelah itu kita lihat type datanya, jika type datanya bukan numerik/float maka perlu diganti type datanya

In [6]:
df_copy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   sepal_length  150 non-null    float64
 1   sepal_width   150 non-null    float64
 2   petal_length  150 non-null    float64
 3   petal_width   150 non-null    float64
 4   species       150 non-null    object 
dtypes: float64(4), object(1)
memory usage: 6.0+ KB


7. Mengubah semuanya tipe data pada fitur menjadi tipe data numerik

In [7]:
# mengubah fitur sepal_length menjadi tipe data float
df_copy['sepal_length'] = df_copy['sepal_length'].astype(float)

# mengubah fitur species menjadi tipe data int yang awalnya categorical
le = LabelEncoder()
df_copy['species'] = le.fit_transform(df_copy['species'])

8. Cek kembali tipe data pada semua fitur

In [8]:
df_copy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   sepal_length  150 non-null    float64
 1   sepal_width   150 non-null    float64
 2   petal_length  150 non-null    float64
 3   petal_width   150 non-null    float64
 4   species       150 non-null    int64  
dtypes: float64(4), int64(1)
memory usage: 6.0 KB


9. Mengisi nilai yang hilang/kosong. Disini kita menggunakan mean, mean tidak selalu menjadi nilai yang baik untuk mengisi nilai yang hilang. Kita juga bisa menggunakan rumus untuk mengisinya.

In [9]:
df_copy = df_copy.fillna(df_copy.mean(numeric_only=True))

10. Cek apakah masih ada data yang kosong

In [10]:
df_copy.isnull().sum()

sepal_length    0
sepal_width     0
petal_length    0
petal_width     0
species         0
dtype: int64

11. Sekarang coba kita tampilkan data yang terbaru

In [11]:
df_copy.head(jumlah_baris)

sepal_length  sepal_width  petal_length  petal_width  species
0             5.1          3.5           1.4          0.2        0
1             4.9          3.0           1.4          0.2        0
2             4.7          3.2           1.3          0.2        0
3             4.6          3.1           1.5          0.2        0
4             5.0          3.6           1.4          0.2        0
..            ...          ...           ...          ...      ...
145           6.7          3.0           5.2          2.3        2
146           6.3          2.5           5.0          1.9        2
147           6.5          3.0           5.2          2.0        2
148           6.2          3.4           5.4          2.3        2
149           5.9          3.0           5.1          1.8        2

[150 rows x 5 columns]

12. Untuk melakukan normalisasi data, metode yang digunakan adalah metode min-max. Metode ini mengubah setiap nilai dalam dataset ke dalam rentang antara 0 hingga 1 dengan menggunakan rumus: (x - min) / (max - min). Hal ini dilakukan untuk setiap kolom kecuali kolom target (kolom species). Terdapat juga metode lain seperti simple feature scaling atau z-score yang dapat digunakan untuk normalisasi data, namun dalam konteks ini, kita akan menggunakan metode min-max untuk memastikan bahwa setiap nilai dalam dataset memiliki skala yang serupa dan tidak mendistorsi perbandingan antar variabel.

In [12]:
species_column = df_copy['species']
df_copy = df_copy[['sepal_length', 'sepal_width', 'petal_length', 'petal_width']].copy()
df_copy = (df_copy - df_copy.min())/(df_copy.max() - df_copy.min())
df_copy['species'] = species_column.copy()
df_copy.head(jumlah_baris)

sepal_length  sepal_width  petal_length  petal_width  species
0        0.222222     0.625000      0.067797     0.041667        0
1        0.166667     0.416667      0.067797     0.041667        0
2        0.111111     0.500000      0.050847     0.041667        0
3        0.083333     0.458333      0.084746     0.041667        0
4        0.194444     0.666667      0.067797     0.041667        0
..            ...          ...           ...          ...      ...
145      0.666667     0.416667      0.711864     0.916667        2
146      0.555556     0.208333      0.677966     0.750000        2
147      0.611111     0.416667      0.711864     0.791667        2
148      0.527778     0.583333      0.745763     0.916667        2
149      0.444444     0.416667      0.694915     0.708333        2

[150 rows x 5 columns]

13. Langkah terakhir dalam proses ini adalah menginisialisasi objek KNeighborsRegressor, yang merupakan algoritma K-Nearest Neighbors untuk melakukan regresi. Setelah itu, kita melakukan randomisasi indeks dari dataset menggunakan np.random.permutation() dengan seed 1 untuk memastikan hasil yang konsisten.

> Selanjutnya, dataset dibagi menjadi dua bagian: 75% untuk training dan 25% untuk testing. Bagian training disimpan dalam variabel training_data, sementara bagian testing disimpan dalam variabel testing_data.

>Setelah itu, model dilatih menggunakan data training dengan memanggil method fit() pada objek KNeighborsRegressor. Argumen input yang diberikan adalah fitur-fitur yang digunakan (sepal_length, sepal_width, petal_length, dan petal_width) dan label (species).

>Kemudian, prediksi dilakukan pada data testing dengan memanggil method predict() pada objek KNeighborsRegressor. Argumen input yang diberikan adalah fitur-fitur dari data testing yang ingin diprediksi.

>Selanjutnya, akurasi model dihitung dengan mengukur persentase data testing yang berhasil diprediksi dengan menggunakan threshold default (biasanya 0.5).

>Terakhir, hasil akurasi model dicetak ke layar.

>Dengan langkah-langkah ini, kita dapat menggunakan algoritma K-Nearest Neighbors untuk melakukan regresi pada dataset yang telah diolah dan memperoleh hasil akurasi model sebagai metrik evaluasi performa.

In [ ]:
knn = KNeighborsRegressor()

# before we split, make sure you randomize the dataset
np.random.seed(1)
index_shuffled = np.random.permutation(df_copy.index)
shuffled_car_data = df_copy.loc[index_shuffled]

# We will use 75% of the data for training and the remaining 25% for testing
train_index = int(shuffled_car_data.shape[0] * 0.75)
training_data = shuffled_car_data.iloc[0:train_index]
testing_data = shuffled_car_data.iloc[train_index:]

# fit the model with the training data
knn.fit(training_data[['sepal_length', 'sepal_width', 'petal_length', 'petal_width']], training_data['species'])

# predict the species of the testing data
prediction = knn.predict(testing_data[['sepal_length', 'sepal_width', 'petal_length', 'petal_width']])

# calculate the accuracy of the model
accuracy = np.mean(np.abs(prediction - testing_data['species']) <= 0.5)

print(f"Akurasi model KNN: {accuracy:.2%}")


Akurasi model KNN: 92.11%


## Kesimpulan

Berdasarkan hasil diatas, tingkat akurasi yang diperoleh dari model KNN pada data iris menggunakan beberapa fitur ('sepal_length', 'sepal_width', 'petal_length', 'petal_width') adalah sebesar 92.11%